test-d/last-array-element.ts

```diff
@@ -0,0 +1,8 @@
+import {expectType} from 'tsd';
+import {LastArrayElement} from '../ts41';
+
+declare function lastOf<V extends any[]>(array: V): LastArrayElement<V>;
+const array = ['foo', 2];
+
+expectType<number>(lastOf(array));
+expectType<string>(lastOf(array.reverse()));
```

Comment:
I'm a bit surprised these tests are passing? Doesn't `array` have type `(string | number)` rather than `[string, number]`? Plus, does `.reverse()` actually reverse tuples in type-land?

Reply:
It's not a tuple, it's an array. Also the entire point of `LastArrayElement` is to get te actual type of the last element. Yes, the array's type is `(string | number)[]` but the type of `array[array.length - 1]` is what we want to extract here, which in this case is `number`.

Reply:
Right, but isn't `LastArrayElement<(string | number)[]>` equal to `string | number`? This test suggests it's `string` which seems odd to me. Not necessarily something wrong with the type itself, but the test (and maybe even `expectType`)

Reply:
Okay, so you inspired to take a more thourough look at this and my VS Code's TS Server seems to think that `lastOf(array)` returns `never`. That would indeed indicate that `expectType` isn't working as intended I think. I need some more time on this type then.

Reply:
It looks like `LastArrayElement` works only on specifically typed arrays:

```TS
const array: ['foo', 2, 'bar'] = ['foo', 2, 'bar'];
lastOf(array); // -> is of type 'bar'

const integers: [0, 1, 2] = [0, 1, 2];
lastOf(array); // -> is of type 2

const elements: [HTMLDivElement, HTMLSpanElement] = [querySelector('div'), querySelector('span')];
lastOf(elements); // -> is of type HTMLSpanElement
```

Otherwise it just returns `never`. Not that helpful of a type to be honest, but necessary for `QueryResult`. I think we might reconsider if it should be exposed.

Reply:
Makes sense. I guess this is another instance of the tsd bug mentioned in https://github.com/sindresorhus/type-fest/pull/153#discussion_r530733893

But also the LastArrayElement type could be improved to fix - maybe instead of the last `: never` it should be `: V[number]`. Then this test would have caught it.

I agree that it simplifies things to not expose the type at all, though.

Generally specifically typed arrays are called tuples, and the slightly easier way to declare them is to use `as const`:

```typescript
const array = ['foo', 2] as const
```

But yeah generally people have arrays, not tuples, available. So this test is probably a bit unrealistic.